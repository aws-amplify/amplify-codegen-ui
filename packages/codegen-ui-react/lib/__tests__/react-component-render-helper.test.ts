/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import {
  RelationalOperator,
  ConditionalStudioComponentProperty,
  StudioComponentProperty,
  ComponentMetadata,
  ConcatenatedStudioComponentProperty,
} from '@aws-amplify/codegen-ui';
import {
  getFixedComponentPropValueExpression,
  getComponentPropName,
  isFixedPropertyWithValue,
  isBoundProperty,
  isCollectionItemBoundProperty,
  isConcatenatedProperty,
  isConditionalProperty,
  buildFixedJsxExpression,
  isDefaultValueOnly,
  getSyntaxKindToken,
  buildChildElement,
  buildConditionalExpression,
  hasChildrenProp,
  buildConcatExpression,
  parseNumberOperand,
} from '../react-component-render-helper';

import { assertASTMatchesSnapshot } from './__utils__';

function buildConditionalWithOperand(operand: string, type?: string) {
  const conditional: ConditionalStudioComponentProperty = {
    condition: {
      property: 'user',
      field: 'age',
      operator: 'gt',
      operand,
      then: {
        value: 'Vote',
      },
      else: {
        value: 'Sorry you cannot vote',
      },
    },
  };

  if (type) {
    conditional.condition.operandType = type;
  }

  return conditional;
}

function buildEmptyComponentMetadata(): ComponentMetadata {
  return {
    hasAuthBindings: false,
    requiredDataModels: [],
    stateReferences: [],
    componentNameToTypeMap: {},
  };
}

describe('react-component-render-helper', () => {
  test('getFixedComponentPropValueExpression', () => {
    const value = 'testValue';
    expect(getFixedComponentPropValueExpression({ value }).text).toEqual(value);
  });

  describe('getComponentPropName', () => {
    test('with name', () => {
      const name = 'ComponentName';
      expect(getComponentPropName(name)).toEqual(`${name}Props`);
    });
    test('without name', () => {
      expect(getComponentPropName()).toEqual(`ComponentWithoutNameProps`);
    });
  });

  describe('property type checkers', () => {
    const propertyTypes: { [propertyType: string]: { checker: Function; property: StudioComponentProperty } } = {
      ConcatenatedStudioComponentProperty: { checker: isConcatenatedProperty, property: { concat: [] } },
      ConditionalStudioComponentProperty: {
        checker: isConditionalProperty,
        property: buildConditionalWithOperand('18'),
      },
      FixedStudioComponentProperty: { checker: isFixedPropertyWithValue, property: { value: 'testValue' } },
      BoundStudioComponentProperty: {
        checker: isBoundProperty,
        property: { bindingProperties: { property: 'testBinding' } },
      },
      CollectionStudioComponentProperty: {
        checker: isCollectionItemBoundProperty,
        property: { collectionBindingProperties: { property: 'testCollectionBinding' } },
      },
    };

    Object.keys(propertyTypes).forEach((propertyType) => {
      describe(propertyType, () => {
        const {
          [propertyType]: { checker, property },
          ...otherProperties
        } = propertyTypes;
        test(`is ${propertyType}`, () => {
          expect(checker(property)).toBeTruthy();
        });

        Object.entries(otherProperties).forEach(([otherPropertyType, { property: otherProperty }]) => {
          test(`${otherPropertyType} is not ${propertyType}`, () => {
            expect(checker(otherProperty)).toBeFalsy();
          });
        });
      });
    });
  });

  describe('buildFixedJsxExpression', () => {
    test('string', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: 'some text' }));
    });

    test('number', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: 400 }));
    });

    test('string wrapped number', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: '400' }));
    });

    test('parsed number', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: '400', type: 'Number' }));
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: '400', type: 'number' }));
    });

    test('boolean', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: true }));
    });

    test('string wrapped boolean', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: 'true' }));
    });

    test('parsed boolean', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: 'true', type: 'Boolean' }));
    });

    test('string wrapped array', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: '[1,2,3]' }));
    });

    test('parsed array', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: '[1,2,3]', type: 'Object' }));
    });

    test('string wrapped object', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: '{"transponder": "rocinante"}' }));
    });

    test('parsed object', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: '{"transponder": "rocinante"}', type: 'Object' }));
    });

    test('string wrapped null', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: 'null' }));
    });

    test('parsed null', () => {
      assertASTMatchesSnapshot(buildFixedJsxExpression({ value: 'null', type: 'Object' }));
    });

    test('type mismatch error', () => {
      expect(() => buildFixedJsxExpression({ value: 'true', type: 'number' })).toThrowErrorMatchingSnapshot();
    });

    test('json parse error', () => {
      expect(() => buildFixedJsxExpression({ value: '⭐', type: 'number' })).toThrowErrorMatchingSnapshot();
    });

    test('unsupported type', () => {
      expect(() =>
        buildFixedJsxExpression({ value: new Date('December 17, 1995 03:24:00'), type: 'unsupported' }),
      ).toThrowError();
      // Not using a snapshot here since the error message changes based on timezone
    });
  });

  describe('isDefaultValueOnly', () => {
    test('defaultValue and not bound property', () => {
      const prop = {
        defaultValue: 'default',
        value: 'something',
      };
      expect(isDefaultValueOnly(prop)).toBeTruthy();
    });

    test('not defaultValue', () => {
      const prop = {
        collectionBindingProperties: {
          property: 'prop',
        },
      };
      expect(isDefaultValueOnly(prop)).toBeFalsy();
    });
  });

  describe('getSyntaxKindToken', () => {
    test('eq', () => {
      expect(getSyntaxKindToken('eq')).toMatchSnapshot();
    });

    test('ne', () => {
      expect(getSyntaxKindToken('ne')).toMatchSnapshot();
    });

    test('le', () => {
      expect(getSyntaxKindToken('le')).toMatchSnapshot();
    });

    test('lt', () => {
      expect(getSyntaxKindToken('lt')).toMatchSnapshot();
    });

    test('ge', () => {
      expect(getSyntaxKindToken('ge')).toMatchSnapshot();
    });

    test('gt', () => {
      expect(getSyntaxKindToken('gt')).toMatchSnapshot();
    });
  });

  describe('buildChildElement', () => {
    test('no prop', () => {
      expect(buildChildElement(buildEmptyComponentMetadata())).toBeUndefined();
    });

    test('fixed property', () => {
      const prop = { value: 'foo' };
      expect(buildChildElement(buildEmptyComponentMetadata(), prop)).toMatchSnapshot();
    });

    test('bound property', () => {
      const prop = { bindingProperties: { property: 'prop' } };
      expect(buildChildElement(buildEmptyComponentMetadata(), prop)).toMatchSnapshot();
    });

    test('collection bound property', () => {
      const prop = { collectionBindingProperties: { property: 'prop' } };
      expect(buildChildElement(buildEmptyComponentMetadata(), prop)).toMatchSnapshot();
    });

    test('concatenated property', () => {
      const prop = { concat: [] };
      expect(buildChildElement(buildEmptyComponentMetadata(), prop)).toMatchSnapshot();
    });

    test('conditonal property', () => {
      const prop = {
        condition: {
          property: 'prop',
          operator: 'eq' as RelationalOperator,
          operand: 0,
          then: { value: 'foo' },
          else: { value: 'bar' },
        },
      };
      expect(buildChildElement(buildEmptyComponentMetadata(), prop)).toMatchSnapshot();
    });
  });

  describe('buildContionalExpression', () => {
    test('operandType exists', () => {
      const numberExpression = buildConditionalExpression(
        buildEmptyComponentMetadata(),
        buildConditionalWithOperand('18', 'number'),
      );
      assertASTMatchesSnapshot(numberExpression);
      const booleanExpression = buildConditionalExpression(
        buildEmptyComponentMetadata(),
        buildConditionalWithOperand('true', 'boolean'),
      );
      assertASTMatchesSnapshot(booleanExpression);
      const stringExpression = buildConditionalExpression(
        buildEmptyComponentMetadata(),
        buildConditionalWithOperand('true', 'string'),
      );
      assertASTMatchesSnapshot(stringExpression);
    });

    test('operandType does not exist', () => {
      const numberExpression = buildConditionalExpression(
        buildEmptyComponentMetadata(),
        buildConditionalWithOperand('18'),
      );
      assertASTMatchesSnapshot(numberExpression);
      const booleanExpression = buildConditionalExpression(
        buildEmptyComponentMetadata(),
        buildConditionalWithOperand('true'),
      );
      assertASTMatchesSnapshot(booleanExpression);
      const stringExpression = buildConditionalExpression(
        buildEmptyComponentMetadata(),
        buildConditionalWithOperand('dlo'),
      );
      assertASTMatchesSnapshot(stringExpression);
    });

    test('operand and operandType mismatch', () => {
      expect(() =>
        buildConditionalExpression(buildEmptyComponentMetadata(), buildConditionalWithOperand('18', 'boolean')),
      ).toThrow('Parsed value 18 and type boolean mismatch');
    });
  });

  describe('hasChildrenProp', () => {
    test('returns true if children property exists', () => {
      expect(
        hasChildrenProp({ width: { value: '10px' }, children: { bindingProperties: { property: 'mySlot' } } }),
      ).toBe(true);
    });

    test('returns false if children property does not exist', () => {
      expect(
        hasChildrenProp({ width: { value: '10px' }, height: { bindingProperties: { property: 'myHeight' } } }),
      ).toBe(false);
    });
  });
  describe('buildConcatExpression', () => {
    test('should build concat with userAttribute', () => {
      const concatProp: ConcatenatedStudioComponentProperty = {
        concat: [{ userAttribute: 'email' }, { value: ', welcome!', type: 'string' }],
      };

      const exp = buildConcatExpression(concatProp);
      assertASTMatchesSnapshot(exp);
    });
  });

  describe('parseNumberOperand', () => {
    test('should parse int if field data type is Int', () => {
      expect(parseNumberOperand('10', { dataType: 'Int', readOnly: false, required: false, isArray: false })).toBe(10);
    });
    test('should parse int if field data type is Int', () => {
      expect(parseNumberOperand('10.01', { dataType: 'Float', readOnly: false, required: false, isArray: false })).toBe(
        10.01,
      );
    });
  });
});
