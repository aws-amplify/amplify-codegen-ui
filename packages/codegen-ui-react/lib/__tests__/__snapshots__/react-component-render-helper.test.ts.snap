// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-component-render-helper buildChildElement bound property 1`] = `
NodeObject {
  "dotDotDotToken": undefined,
  "end": -1,
  "expression": IdentifierObject {
    "end": -1,
    "escapedText": "prop",
    "flags": 8,
    "kind": 79,
    "modifierFlagsCache": 0,
    "originalKeywordKind": undefined,
    "parent": undefined,
    "pos": -1,
    "transformFlags": 0,
  },
  "flags": 8,
  "kind": 286,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 2,
}
`;

exports[`react-component-render-helper buildChildElement collection bound property 1`] = `
NodeObject {
  "dotDotDotToken": undefined,
  "end": -1,
  "expression": IdentifierObject {
    "end": -1,
    "escapedText": "item",
    "flags": 8,
    "kind": 79,
    "modifierFlagsCache": 0,
    "originalKeywordKind": undefined,
    "parent": undefined,
    "pos": -1,
    "transformFlags": 0,
  },
  "flags": 8,
  "kind": 286,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 2,
}
`;

exports[`react-component-render-helper buildChildElement concatenated property 1`] = `
NodeObject {
  "dotDotDotToken": undefined,
  "end": -1,
  "expression": NodeObject {
    "end": -1,
    "flags": 8,
    "head": TokenObject {
      "end": -1,
      "flags": 8,
      "kind": 15,
      "modifierFlagsCache": 0,
      "parent": undefined,
      "pos": -1,
      "rawText": "",
      "templateFlags": 0,
      "text": "",
      "transformFlags": 512,
    },
    "kind": 221,
    "modifierFlagsCache": 0,
    "parent": undefined,
    "pos": -1,
    "templateSpans": Array [],
    "transformFlags": 512,
  },
  "flags": 8,
  "kind": 286,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 514,
}
`;

exports[`react-component-render-helper buildChildElement conditonal property 1`] = `
NodeObject {
  "dotDotDotToken": undefined,
  "end": -1,
  "expression": NodeObject {
    "colonToken": TokenObject {
      "end": -1,
      "flags": 8,
      "kind": 58,
      "modifierFlagsCache": 0,
      "parent": undefined,
      "pos": -1,
      "transformFlags": 0,
    },
    "condition": NodeObject {
      "end": -1,
      "flags": 8,
      "kind": 219,
      "left": IdentifierObject {
        "end": -1,
        "escapedText": "prop",
        "flags": 8,
        "kind": 79,
        "modifierFlagsCache": 0,
        "originalKeywordKind": undefined,
        "parent": undefined,
        "pos": -1,
        "transformFlags": 0,
      },
      "modifierFlagsCache": 0,
      "operatorToken": TokenObject {
        "end": -1,
        "flags": 8,
        "kind": 34,
        "modifierFlagsCache": 0,
        "parent": undefined,
        "pos": -1,
        "transformFlags": 0,
      },
      "parent": undefined,
      "pos": -1,
      "right": TokenObject {
        "end": -1,
        "flags": 8,
        "kind": 8,
        "modifierFlagsCache": 0,
        "numericLiteralFlags": 0,
        "parent": undefined,
        "pos": -1,
        "text": "0",
        "transformFlags": 0,
      },
      "transformFlags": 0,
    },
    "end": -1,
    "flags": 8,
    "kind": 220,
    "modifierFlagsCache": 0,
    "parent": undefined,
    "pos": -1,
    "questionToken": TokenObject {
      "end": -1,
      "flags": 8,
      "kind": 57,
      "modifierFlagsCache": 0,
      "parent": undefined,
      "pos": -1,
      "transformFlags": 0,
    },
    "transformFlags": 0,
    "whenFalse": TokenObject {
      "end": -1,
      "flags": 8,
      "hasExtendedUnicodeEscape": undefined,
      "kind": 10,
      "modifierFlagsCache": 0,
      "parent": undefined,
      "pos": -1,
      "singleQuote": undefined,
      "text": "bar",
      "transformFlags": 0,
    },
    "whenTrue": TokenObject {
      "end": -1,
      "flags": 8,
      "hasExtendedUnicodeEscape": undefined,
      "kind": 10,
      "modifierFlagsCache": 0,
      "parent": undefined,
      "pos": -1,
      "singleQuote": undefined,
      "text": "foo",
      "transformFlags": 0,
    },
  },
  "flags": 8,
  "kind": 286,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 2,
}
`;

exports[`react-component-render-helper buildChildElement fixed property 1`] = `
NodeObject {
  "dotDotDotToken": undefined,
  "end": -1,
  "expression": TokenObject {
    "end": -1,
    "flags": 8,
    "hasExtendedUnicodeEscape": undefined,
    "kind": 10,
    "modifierFlagsCache": 0,
    "parent": undefined,
    "pos": -1,
    "singleQuote": undefined,
    "text": "foo",
    "transformFlags": 0,
  },
  "flags": 8,
  "kind": 286,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 2,
}
`;

exports[`react-component-render-helper buildConcatExpression should build concat with userAttribute 1`] = `"\`\${authAttributes[\\"email\\"]}\${\\", welcome!\\"}\`"`;

exports[`react-component-render-helper buildContionalExpression operandType does not exist 1`] = `"user?.age > \\"18\\" ? \\"Vote\\" : \\"Sorry you cannot vote\\""`;

exports[`react-component-render-helper buildContionalExpression operandType does not exist 2`] = `"user?.age > \\"true\\" ? \\"Vote\\" : \\"Sorry you cannot vote\\""`;

exports[`react-component-render-helper buildContionalExpression operandType does not exist 3`] = `"user?.age > \\"dlo\\" ? \\"Vote\\" : \\"Sorry you cannot vote\\""`;

exports[`react-component-render-helper buildContionalExpression operandType exists 1`] = `"user?.age > 18 ? \\"Vote\\" : \\"Sorry you cannot vote\\""`;

exports[`react-component-render-helper buildContionalExpression operandType exists 2`] = `"user?.age > true ? \\"Vote\\" : \\"Sorry you cannot vote\\""`;

exports[`react-component-render-helper buildContionalExpression operandType exists 3`] = `"user?.age > \\"true\\" ? \\"Vote\\" : \\"Sorry you cannot vote\\""`;

exports[`react-component-render-helper buildFixedJsxExpression boolean 1`] = `"{true}"`;

exports[`react-component-render-helper buildFixedJsxExpression json parse error 1`] = `"Failed to parse value \\"‚≠ê\\""`;

exports[`react-component-render-helper buildFixedJsxExpression number 1`] = `"{400}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed array 1`] = `"{[1, 2, 3]}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed boolean 1`] = `"{true}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed null 1`] = `"{null}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed number 1`] = `"{400}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed number 2`] = `"{400}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed object 1`] = `"{{ \\"transponder\\": \\"rocinante\\" }}"`;

exports[`react-component-render-helper buildFixedJsxExpression string 1`] = `"\\"some text\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped array 1`] = `"\\"[1,2,3]\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped boolean 1`] = `"\\"true\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped null 1`] = `"\\"null\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped number 1`] = `"\\"400\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped object 1`] = `"\\"{\\\\\\"transponder\\\\\\": \\\\\\"rocinante\\\\\\"}\\""`;

exports[`react-component-render-helper buildFixedJsxExpression type mismatch error 1`] = `"Parsed value type \\"boolean\\" and specified type \\"number\\" mismatch"`;

exports[`react-component-render-helper getSyntaxKindToken eq 1`] = `
TokenObject {
  "end": -1,
  "flags": 8,
  "kind": 34,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 0,
}
`;

exports[`react-component-render-helper getSyntaxKindToken ge 1`] = `
TokenObject {
  "end": -1,
  "flags": 8,
  "kind": 33,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 0,
}
`;

exports[`react-component-render-helper getSyntaxKindToken gt 1`] = `
TokenObject {
  "end": -1,
  "flags": 8,
  "kind": 31,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 0,
}
`;

exports[`react-component-render-helper getSyntaxKindToken le 1`] = `
TokenObject {
  "end": -1,
  "flags": 8,
  "kind": 32,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 0,
}
`;

exports[`react-component-render-helper getSyntaxKindToken lt 1`] = `
TokenObject {
  "end": -1,
  "flags": 8,
  "kind": 29,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 0,
}
`;

exports[`react-component-render-helper getSyntaxKindToken ne 1`] = `
TokenObject {
  "end": -1,
  "flags": 8,
  "kind": 35,
  "modifierFlagsCache": 0,
  "parent": undefined,
  "pos": -1,
  "transformFlags": 0,
}
`;
